pkg/calculator/experimentcalc/experiment_calculator.go

```diff
@@ -0,0 +1,650 @@
+			gr.GoalId = experiment.GoalId
+			e.appendVariationResult(ctx, timestamp, goalResult, gr.VariationResults)
+		}
+		experimentResult.GoalResults = append(experimentResult.GoalResults, goalResult)
+	}
+
+	return experimentResult
+}
+
+func (e ExperimentCalculator) listEnvironments(
+	ctx context.Context,
+) ([]*environment.Environment, error) {
+	listEnvironmentsRequest := environment.ListEnvironmentsRequest{
+		PageSize: 0,
+		Cursor:   "",
+	}
+	resp, err := e.environmentClient.ListEnvironments(ctx, &listEnvironmentsRequest)
+	if err != nil {
+		return nil, err
+	}
+	return resp.Environments, err
+}
+
+func (e ExperimentCalculator) listExperiments(
+	ctx context.Context,
+	namespace string,
+) ([]*experiment.Experiment, error) {
+	req := &experiment.ListExperimentsRequest{
+		From:                 time.Now().Add(-2 * 24 * time.Hour).Unix(),
+		PageSize:             0,
+		Cursor:               "",
+		EnvironmentNamespace: namespace,
+		Statuses: []experiment.Experiment_Status{
+			experiment.Experiment_WAITING,
+			experiment.Experiment_RUNNING,
+		},
+	}
+	resp, err := e.experimentClient.ListExperiments(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+	return resp.Experiments, nil
+}
+
+func (e ExperimentCalculator) getEvaluationCount(
+	ctx context.Context,
+	req *eventcounter.GetExperimentEvaluationCountRequest,
+) (map[string]*eventcounter.VariationCount, error) {
+	variationCounts := make(map[string]*eventcounter.VariationCount)
+	evaluationCount, err := e.eventCounterClient.GetExperimentEvaluationCount(ctx, req)
+	if err != nil {
+		return variationCounts, err
+	}
+	for _, variationCount := range evaluationCount.VariationCounts {
+		variationCounts[variationCount.VariationId] = variationCount
+	}
+	return variationCounts, nil
+}
+
+func (e ExperimentCalculator) getGoalCount(
+	ctx context.Context,
+	req *eventcounter.GetExperimentGoalCountRequest,
+) (map[string]*eventcounter.VariationCount, error) {
+	variationCounts := make(map[string]*eventcounter.VariationCount)
+	goalCount, err := e.eventCounterClient.GetExperimentGoalCount(ctx, req)
+	if err != nil {
+		return variationCounts, err
+	}
+	for _, variationCount := range goalCount.VariationCounts {
+		variationCounts[variationCount.VariationId] = variationCount
+	}
+	return variationCounts, nil
+}
+
+func (e ExperimentCalculator) calcGoalResult(
+	ctx context.Context,
+	evalVariationCounts,
+	goalVariationCounts map[string]*eventcounter.VariationCount,
+	baseVid string,
+) *eventcounter.GoalResult {
+	goalResult := &eventcounter.GoalResult{}
+	length := len(goalVariationCounts)
+	vids := make([]string, 0, length)
+	goalUc, evalUc := make([]int64, 0, length), make([]int64, 0, length)
+	vrs := make(map[string]*eventcounter.VariationResult, length)
+	valueMeans, valueVars := make([]float64, 0, length), make([]float64, 0, length)
+	baselineIdx, loopIdx := 0, 0
+
+	for vid, goalVariationCount := range goalVariationCounts {
+		if _, ok := evalVariationCounts[vid]; !ok {
+			return nil
```

Comment:
The return value of this function is to be accessed in [the next line](https://github.com/bucketeer-io/bucketeer/pull/395/files#diff-6b2e151d6e7453bb8a10c3b450be4ba51fc681fe35b1b485d43d452d0deb19b5R195-R196), so returning nil would result in a nil pointer exception.

Reply:
Of cause. Thanks for pointing this out!

Reply:
Fixed